### PR TITLE
Emitter: handle packet send fungible token

### PR DESCRIPTION
**Pull request by Benzbeeb**
Fixed: #XXXX

## Please ensure the following requirements are met before submitting a pull request:

- [ ] The pull request is targeted against the correct target branch
- [ ] The pull request is linked to an issue with appropriate discussion and an accepted design OR is linked to a spec that describes the work.
- [ ] The pull request includes a description of the implementation/work done in detail.
- [ ] The pull request includes any and all appropriate unit/integration tests
- [ ] You have added a relevant changelog entry to `CHANGELOG_UNRELEASED.md`
- [ ] You have re-reviewed the files affected by the pull request (e.g. using the `Files changed` tab in the Github PR explorer)

### CODE REVIEW

File: hooks/emitter/channel.go [hunk 1]
```diff
@@ -1,30 +1,47 @@
...
-		"src_channel":  msg.Packet.SourceChannel,
-		"src_port":     msg.Packet.SourcePort,
-		"sequence":     msg.Packet.Sequence,
-		"dst_channel":  msg.Packet.DestinationChannel,
-		"dst_port":     msg.Packet.DestinationPort,
+func (h *Hook) handleFungibleTokenPacket(
+	ctx sdk.Context, txHash []byte, msg *types.MsgRecvPacket, packet common.JsDict, evMap common.EvMap, extra common.JsDict,
+) bool {
+	var data ibcxfertypes.FungibleTokenPacketData
+	err := oracletypes.ModuleCdc.UnmarshalJSON(msg.Packet.Data, &data)
```
Comment (taobun) [inline hooks/emitter/channel.go:17]: ibcxfertypes.ModuleCdc

File: hooks/emitter/transfer.go [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	channeltypes "github.com/cosmos/cosmos-sdk/x/ibc/core/04-channel/types"
+
+	"github.com/bandprotocol/chain/hooks/common"
+)
+
+// handleMsgTransfer implements emitter handler for msgTransfer.
+func (h *Hook) handleMsgTransfer(ctx sdk.Context, msg *types.MsgTransfer, evMap common.EvMap) {
+	if events, ok := evMap[channeltypes.EventTypeSendPacket+"."+channeltypes.AttributeKeyData]; ok {
+		var data ibcxfertypes.FungibleTokenPacketData
+		err := h.cdc.UnmarshalJSON([]byte(events[0]), &data)
```
Comment (taobun) [inline hooks/emitter/transfer.go:16]: ibcxfertypes.ModuleCdc?

File: hooks/emitter/channel.go [hunk 3]
```diff
@@ -1,30 +1,47 @@
...
-) {
-	packet := common.JsDict{
-		"is_incoming":  true,
-		"block_height": ctx.BlockHeight(),
-		"src_channel":  msg.Packet.SourceChannel,
-		"src_port":     msg.Packet.SourcePort,
-		"sequence":     msg.Packet.Sequence,
-		"dst_channel":  msg.Packet.DestinationChannel,
-		"dst_port":     msg.Packet.DestinationPort,
+func (h *Hook) handleFungibleTokenPacket(
```
Comment (taobun) [inline hooks/emitter/channel.go:13]: change name to `extractFungibleTokenPacket(ctx, packetData []byte,evMap common.EvMap, packet)`

File: hooks/emitter/transfer.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+		if err == nil {
+			packet := common.JsDict{
+				"is_incoming":  false,
+				"block_height": ctx.BlockHeight(),
+				"src_channel":  evMap[channeltypes.EventTypeSendPacket+"."+channeltypes.AttributeKeySrcChannel][0],
+				"src_port":     evMap[channeltypes.EventTypeSendPacket+"."+channeltypes.AttributeKeySrcPort][0],
+				"sequence":     common.Atoui(evMap[channeltypes.EventTypeSendPacket+"."+channeltypes.AttributeKeySequence][0]),
+				"dst_channel":  evMap[channeltypes.EventTypeSendPacket+"."+channeltypes.AttributeKeyDstChannel][0],
+				"dst_port":     evMap[channeltypes.EventTypeSendPacket+"."+channeltypes.AttributeKeyDstPort][0],
+			}
```
Comment (taobun) [inline hooks/emitter/transfer.go:26]: This code look duplicate many times can refractor to one function?

**Review by taobun — CHANGES_REQUESTED** [review]

File: hooks/emitter/channel.go [hunk 5]
```diff
@@ -90,23 +107,36 @@ func (h *Hook) handleMsgRecvPacket(
...
+		return
 	}
-	// TODO: Check on other packet
+}
+
+func (h *Hook) handleOracleResponsePacket(
+	ctx sdk.Context, packet common.JsDict, evMap common.EvMap,
+) bool {
 	var data oracletypes.OracleResponsePacketData
 	err := h.cdc.UnmarshalJSON([]byte(evMap[types.EventTypeSendPacket+"."+types.AttributeKeyData][0]), &data)
```
Comment (songwongtp) [inline hooks/emitter/channel.go:141]: oracletypes.ModuleCdc

File: hooks/emitter/channel.go [hunk 6]
```diff
@@ -1,30 +1,59 @@
...
+		}
+		h.Write("NEW_PACKET", packet)
+		return true
 	}
+	return false
+}
 
-	// TODO: Check on other packet
+func (h *Hook) extractOracleRequestPacket(
+	ctx sdk.Context, txHash []byte, msg *types.MsgRecvPacket, packet common.JsDict, evMap common.EvMap, extra common.JsDict,
```
Comment (taobun) [inline hooks/emitter/channel.go:55]: Change msg -> data of packet ([]byte)

File: hooks/emitter/transfer.go [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+	channeltypes "github.com/cosmos/cosmos-sdk/x/ibc/core/04-channel/types"
+
+	"github.com/bandprotocol/chain/hooks/common"
+)
+
+// handleMsgTransfer implements emitter handler for msgTransfer.
+func (h *Hook) handleMsgTransfer(ctx sdk.Context, msg *types.MsgTransfer, evMap common.EvMap) {
+	if events, ok := evMap[channeltypes.EventTypeSendPacket+"."+channeltypes.AttributeKeyData]; ok {
+		packet := h.getPacket(ctx, evMap, false)
+		var data ibcxfertypes.FungibleTokenPacketData
```
Comment (taobun) [inline hooks/emitter/transfer.go:16]: using `extractOracleRequestPacket`

**Review by taobun — CHANGES_REQUESTED** [review]

File: hooks/emitter/channel.go [hunk 8]
```diff
@@ -1,32 +1,65 @@
...
 
-// handleMsgRequestData implements emitter handler for MsgRequestData.
-func (h *Hook) handleMsgRecvPacket(
-	ctx sdk.Context, txHash []byte, msg *types.MsgRecvPacket, evMap common.EvMap, extra common.JsDict,
-) {
-	packet := common.JsDict{
-		"is_incoming":  true,
+// TODO: update transfer acknowledgement for fungible token packet
+
+func (h *Hook) getPacket(ctx sdk.Context, srcPort string, srcChannel string, sequence uint64, dstPort string, dstChannel string, isIncoming bool) common.JsDict {
```
Comment (taobun) [inline hooks/emitter/channel.go:15]: newPacket

File: hooks/emitter/channel.go [hunk 9]
```diff
@@ -1,32 +1,65 @@
...
+				packet["acknowledgement"] = common.JsDict{
+					"success": true,
+				}
+			} else {
+				packet["acknowledgement"] = common.JsDict{
+					"success": false,
+				}
+			}
+		}
+		h.Write("NEW_PACKET", packet)
```
Comment (taobun) [inline hooks/emitter/channel.go:52]: I think extract only add field to packet and call write on caller.

File: hooks/emitter/channel.go [hunk 10]
```diff
@@ -1,32 +1,65 @@
...
+		"src_channel":  srcChannel,
+		"src_port":     srcPort,
+		"sequence":     sequence,
+		"dst_channel":  dstChannel,
+		"dst_port":     dstPort,
+	}
+}
+
+func (h *Hook) extractFungibleTokenPacket(
+	ctx sdk.Context, dataOfPacket []byte, packet common.JsDict, evMap common.EvMap,
```
Comment (taobun) [inline hooks/emitter/channel.go:28]: reorder to ctx, data, evMap, packet

File: hooks/emitter/channel.go [hunk 11]
```diff
@@ -1,32 +1,65 @@
...
+		}
+		h.Write("NEW_PACKET", packet)
+		return true
 	}
+	return false
+}
 
-	// TODO: Check on other packet
+func (h *Hook) extractOracleRequestPacket(
+	ctx sdk.Context, txHash []byte, signer string, dataOfPacket []byte, packet common.JsDict, evMap common.EvMap, extra common.JsDict,
```
Comment (taobun) [inline hooks/emitter/channel.go:59]: ditto

Comment (taobun) [reply hooks/emitter/channel.go:59]: Remove txHash?

**Review by taobun — CHANGES_REQUESTED** [review]

File: hooks/emitter/channel.go [hunk 12]
```diff
@@ -89,26 +121,38 @@ func (h *Hook) handleMsgRecvPacket(
...
+	packet := newPacket(
+		ctx,
+		msg.Packet.SourcePort,
+		msg.Packet.SourceChannel,
+		msg.Packet.Sequence,
+		msg.Packet.DestinationPort,
+		msg.Packet.DestinationChannel,
+		true,
+	)
+	h.Write("NEW_PACKET", packet)
```
Comment (taobun) [inline hooks/emitter/channel.go:142]: This is your intention to put packet to list first then modify later?

**Review by taobun — APPROVED** [review]

**Review by songwongtp — APPROVED** [review]

**Pull request merged**